internal/configfsm/fsm.go

```diff
@@ -0,0 +1,309 @@
+	ClusterNotReadyToClusterNotReady Trigger = "ClusterNotReady->ClusterNotReady"
+	InProgressToInProgress           Trigger = "InProgress->InProgress"
+	TimedOutToTimedOut               Trigger = "TimedOut->TimedOut"
+	CompletedToCompleted             Trigger = "Completed->Completed"
+	OutOfDateToOutOfDate             Trigger = "OutOfDate->OutOfDate"
+
+	// States
+	Missing         = "Missing"
+	Start           = "Start"
+	ClusterNotReady = "ClusterNotReady"
+	InProgress      = "InProgress"
+	TimedOut        = "TimedOut"
+	Completed       = "Completed"
+	OutOfDate       = "OutOfDate"
+)
+
+// RunFSM Runs the state machine as much as it can by triggering all Self transitions
+func RunFSM(ctx context.Context, fsm *stateless.StateMachine, fsmHelper FsmHelper) (aState any, err error) {
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	switch aState {
+	case Start:
+		err = fsm.Fire(StartToMissing, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Missing:
+		err = fsm.Fire(MissingToMissing, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case ClusterNotReady:
+		err = fsm.Fire(ClusterNotReadyToClusterNotReady, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case InProgress:
+		err = fsm.Fire(InProgressToInProgress, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case TimedOut:
+		err = fsm.Fire(TimedOutToTimedOut, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Completed:
+		err = fsm.Fire(CompletedToCompleted, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case OutOfDate:
+		err = fsm.Fire(OutOfDateToOutOfDate, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	}
+	// Update the current state
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	return aState, nil
+}
+
+// InitFSM Initialize the state machine algorithm
+func InitFSM(state string) (fsm *stateless.StateMachine, err error) {
+	fsm = stateless.NewStateMachine(state)
+	fsm.Configure(Start).Permit(StartToMissing, Missing)
+	fsm.Configure(ClusterNotReady).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering ClusterNotReady")
+			if fsmHelper.IsTimedOut() {
+				return fsm.Fire(ClusterNotReadyToTimedOut, fsmHelper)
+			}
+			if fsmHelper.IsClusterReady() {
+				if fsmHelper.IsNonCompliantPolicyInEnforce() || fsmHelper.IsAllPoliciesCompliant() {
+					return fsm.Fire(ClusterNotReadyToInProgress, fsmHelper)
```

Comment:
Normally, if all the policies are compliant, it should go to `Completed`.
Is the idea here that it will go to `InProgress` first and then to `Completed`?

Reply:
Yes, correct.